### ISSUE TITLE
FFmpeg module filter documentation / Code cleanup

### ISSUE THREAD

**Issue by jpc0** [CasparCG/server#1484]
### Expected behaviour
[https://github.com/CasparCG/server/blob/b0ecbeea48e9a9548199ed1d2cf0d241fa3695fa/src/modules/ffmpeg/producer/av_producer.cpp#L273](https://github.com/CasparCG/server/blob/b0ecbeea48e9a9548199ed1d2cf0d241fa3695fa/src/modules/ffmpeg/producer/av_producer.cpp#L273)
A lot of the filter code exists but it isn't entirely clear why it exists, for instance in [this](https://github.com/CasparCG/server/blob/b0ecbeea48e9a9548199ed1d2cf0d241fa3695fa/src/modules/ffmpeg/producer/av_producer.cpp#L363) loop we check for whether the disposition has been set to default or is unset and these streams are later merged together using an alphamerge. This indicates to me that this is for forced subtitles however should be also be looking for `disposition == AV_DISPOSITION_FORCED` ? This could be a subtle but that has crept in because this should probably be in a separate function or at least commented.

### Current behaviour
A clear and concise description of what the current behaviour is in a comment or in a function name.

### Additional Information
I'm happy to work in this but need to know what is actually happening in the codebase since there may be checks put in place in here to catch edge cases but that hasn't been documented. FFmpeg is a pretty obscure library, any documentation as to why something was done goes a long way, specially when what happens in this codebase diverges from FFmpeg example code.

**Comment by Julusian** [comment]
I know very little about the ffmpeg code we have, it is an area I have been trying to avoid where possible, so I won't be able to provide many answers. Hopefully the git history can provide some insights?

I have a feeling that alphamerge example was done as a way to allow for having a file containing 2 h264 streams in a file, and to use the second as the alpha channel. I guess as an alternative method to the `_A` method we have.

**Comment by jpc0** [comment]
> I have a feeling that alphamerge example was done as a way to allow for having a file containing 2 h264 streams in a file, and to use the second as the alpha channel. I guess as an alternative method to the _A method we have.

That is also possible. Honestly I think there are so many different uses of casparcg out there I'm not sure without documentation that we wouldn't be able to break someone's use case. I mean my primary use case is media playout to OBS personally and for work it is a glorified NDI->SDI converter...

**Comment by jpc0** [comment]
[https://github.com/CasparCG/help/wiki/FFmpeg-Producer#transitions-between-videos](https://github.com/CasparCG/help/wiki/FFmpeg-Producer#transitions-between-videos)
> Transitions Between Videos
> You can play two simultaneous video files from disk to the same consumer during any of the built-in transitions such as mix, wipe, push and slide of a user-selectable number of frames. If your video files have alpha channel (key) those will work as well.

Do you know how this would be invoked from AMCP?

> I have a feeling that alphamerge example was done as a way to allow for having a file containing 2 h264 streams in a file, and to use the second as the alpha channel. I guess as an alternative method to the _A method we have.

I'm thinking that the code [here](https://github.com/CasparCG/server/blob/b0ecbeea48e9a9548199ed1d2cf0d241fa3695fa/src/modules/ffmpeg/producer/av_producer.cpp#L363) from above is to implement this.

**Comment by jpc0** [comment]
[https://github.com/CasparCG/server/pull/696](https://github.com/CasparCG/server/pull/696)

So this is a feature where you can create a network stream with a key and a fill channel which will get merged together into a straight alpha video in CasparCG.

I'm guessing this is in use somewhere since it was brought back in [https://github.com/CasparCG/server/commit/0ceba5a06cf1d376f0bea56560259a264efaccae](https://github.com/CasparCG/server/commit/0ceba5a06cf1d376f0bea56560259a264efaccae) after being removed.

**Comment by Julusian** [comment]
> [CasparCG/help/wiki/FFmpeg-Producer#transitions-between-videos](https://github.com/CasparCG/help/wiki/FFmpeg-Producer#transitions-between-videos)
> 
> > Transitions Between Videos
> > You can play two simultaneous video files from disk to the same consumer during any of the built-in transitions such as mix, wipe, push and slide of a user-selectable number of frames. If your video files have alpha channel (key) those will work as well.
> 
> Do you know how this would be invoked from AMCP?

Those docs are a bit misleading, the transition code is generic and not specific to the ffmpeg producer.
The transition producer is created from amcp at https://github.com/CasparCG/server/blob/master/src/protocol/amcp/AMCPCommandsImpl.cpp#L289-L303
The implementation of these transitions is in https://github.com/CasparCG/server/tree/master/src/core/producer/transition
And has some help https://github.com/CasparCG/server/blob/master/src/core/producer/layer.cpp#L77-L89 to be provided the 'leading_producer' (the one being transition away from)

**Comment by jpc0** [comment]
> Those docs are a bit misleading, the transition code is generic and not specific to the ffmpeg producer.

Makes sense. Thought it might be a little easier to do it somewhere else. Will probably have a think about how to reword that section or just remove it...

**Comment by 5opr4ni** [comment]
> https://github.com/CasparCG/help/wiki/FFmpeg-Producer#transitions-between-videos
> 
> > Transitions Between Videos
> > You can play two simultaneous video files from disk to the same consumer during any of the built-in transitions such as mix, wipe, push and slide of a user-selectable number of frames. If your video files have alpha channel (key) those will work as well.
> 
> Do you know how this would be invoked from AMCP?
> 
> > I have a feeling that alphamerge example was done as a way to allow for having a file containing 2 h264 streams in a file, and to use the second as the alpha channel. I guess as an alternative method to the _A method we have.
> 
> I'm thinking that the code [here](https://github.com/CasparCG/server/blob/b0ecbeea48e9a9548199ed1d2cf0d241fa3695fa/src/modules/ffmpeg/producer/av_producer.cpp#L363) from above is to implement this.

It's made for h264 containing 2 video tracks. One Fill and one Key.

**Comment by 5opr4ni** [comment]
> https://github.com/CasparCG/help/wiki/FFmpeg-Producer#transitions-between-videos
> 
> > Transitions Between Videos
> > You can play two simultaneous video files from disk to the same consumer during any of the built-in transitions such as mix, wipe, push and slide of a user-selectable number of frames. If your video files have alpha channel (key) those will work as well.
> 
> Do you know how this would be invoked from AMCP?
> 
> > I have a feeling that alphamerge example was done as a way to allow for having a file containing 2 h264 streams in a file, and to use the second as the alpha channel. I guess as an alternative method to the _A method we have.
> 
> I'm thinking that the code [here](https://github.com/CasparCG/server/blob/b0ecbeea48e9a9548199ed1d2cf0d241fa3695fa/src/modules/ffmpeg/producer/av_producer.cpp#L363) from above is to implement this.

True! Fill and Key
